fault/tester/abstract_tester.py

```diff
@@ -0,0 +1,231 @@
+import magma as m
+from abc import abstractmethod
+from ..wrapper import CircuitWrapper
+from ..select_path import SelectPath
+from ..wrapped_internal_port import WrappedVerilogInternalPort
+from ..magma_utils import is_recursive_type
+
+
+class AbstractTester:
+    def __init__(self, circuit: m.Circuit, clock: m.Clock = None,
+                 reset: m.Reset = None, poke_delay_default=None,
+                 expect_strict_default=True):
+        """
+        `circuit`: the device under test (a magma circuit)
+        `clock`: optional, a port from `circuit` corresponding to the clock
+        `reset`: optional, a port from `circuit` corresponding to the reset
+        `expect_strict_default`: if True, use strict equality check if
+        not specified by the user.
+        """
+        self._circuit = circuit
+        self.poke_delay_default = poke_delay_default
+        self.expect_strict_default = expect_strict_default
+        self.actions = []
+        if clock is not None and not isinstance(clock, m.Clock):
+            raise TypeError(f"Expected clock port: {clock, type(clock)}")
+        self.clock = clock
+        # Make sure the user has initialized the clock before stepping it
+        # While verilator initializes the clock value to 0, this assumption
+        # does not hold for system verilog, so we log a warning (as to not
+        # break existing TBs)
+        self.clock_initialized = False
+        # Only report once, in case the user calls step with an uninitialized
+        # clock many times
+        self.clock_init_warning_reported = False
+        if reset is not None and not isinstance(reset, m.Reset):
+            raise TypeError(f"Expected reset port: {reset, type(reset)}")
+        self.reset_port = reset
+
+    def get_port_type(self, port):
```

Comment:
needs to be a member function or can be helper?

Reply:
Yes, the children inherit this helper and refer to it